java/ql/src/Security/CWE/CWE-022/TaintedPath.md

```diff
@@ -35,25 +35,50 @@ public void sendUserFile(Socket sock, String user) {
 }
 
 ```
-Simply checking that the path is under a trusted location (such as a known public folder) is not enough, however, since the path could contain relative components such as "..". To fix this, check that it does not contain ".." and starts with the public folder.
+If the input is just supposed to be a file name, you can check that it doesn't contain any path separators or ".." sequences.
 
 
 ```java
 public void sendUserFileGood(Socket sock, String user) {
 	BufferedReader filenameReader = new BufferedReader(
 			new InputStreamReader(sock.getInputStream(), "UTF-8"));
 	String filename = filenameReader.readLine();
-	// GOOD: ensure that the file is in a designated folder in the user's home directory
-	if (!filename.contains("..") && filename.startsWith("/home/" + user + "/public/")) {
-		BufferedReader fileReader = new BufferedReader(new FileReader(filename));
-		String fileLine = fileReader.readLine();
-		while(fileLine != null) {
-			sock.getOutputStream().write(fileLine.getBytes());
-			fileLine = fileReader.readLine();
-		}
+	// GOOD: ensure that the filename has no path separators or parent directory references
+	if (filename.contains("..") || filename.contains("/") || filename.contains("\\")) {
+		throw new IllegalArgumentException("Invalid filename");
+	}
+	BufferedReader fileReader = new BufferedReader(new FileReader(filename));
+	String fileLine = fileReader.readLine();
+	while(fileLine != null) {
+		sock.getOutputStream().write(fileLine.getBytes());
+		fileLine = fileReader.readLine();
 	}	
 }
 
+```
+If the input is supposed to be found within a specific directory, you can check that the resolved path is still contained within that directory.
+
+
+```java
+public void sendUserFileGood(Socket sock, String user) {
+	BufferedReader filenameReader = new BufferedReader(
+			new InputStreamReader(sock.getInputStream(), "UTF-8"));
+	String filename = filenameReader.readLine();
+
+	Path publicFolder = Paths.get("/home/" + user + "/public").normalize().toAbsolutePath();
+	Path filePath = publicFolder.resolve(filename).normalize().toAbsolutePath();
+
+	// GOOD: ensure that the path stays within the public folder
+	if (!filePath.startsWith(publicFolder)) {
```

Comment:
Is it worth refining this check by adding a path separator after `publicFolder`?